lib/util/mutex.js

```diff
@@ -0,0 +1,158 @@
+/*
```

Comment:
Discussion: This class should be exposed in the `kuzzle` package and the plugin context so we can use it in plugins and Kaaf.

It should be rewritten in TS and we need to find a way to get ride of the `kuzzle` constructor argument as well.

But I think we should do that later in a separate PR

Reply:
Ah, I thought about making it available to plugins, but I didn't think that would mean writing it in typescript. This should be converted in another PR though, but yes, I agree with that.

I don't see how we could remove `kuzzle` from the constructor, since we need to emit events to access the cache module. I don't think this is a problem, though. Plugins for instance could have access to a wrapper to this class for instance. Same for kaaf I think.

Reply:
We talked about putting the `kuzzle` instance in the `process` global object so the `ask` and `onAsk` method can be in an independent file and we don't need to pass the kuzzle object in every constructors.  
It will allow to expose the `Mutex` class directly as well;
What do you think?

Reply:
I don't like the idea of adding properties to "process".

I would prefer if we would force the `Kuzzle` object to be a singleton, meaning that we shouldn't be able to instantiate multiple kuzzle objects. If we do so, we could simply add a static getter method returning that singleton.
This means that we could simply do this:

```js
const Kuzzle = require('kuzzle');

const kuzzle = Kuzzle.getInstance();
```

Reply:
Or we could simply expose the `kuzzle` instance directly?

```
const { kuzzle } = require('kuzzle')
```

I'm not sure we can do that thus.

We can discuss it when it comes to expose the Mutex class :+1:

Reply:
We could do what you propose. But, hmm... this would mean we could do this: 

```js
const { Kuzzle, kuzzle } = require('kuzzle');
```

![bakk7](https://user-images.githubusercontent.com/12997967/96149669-dfdf2e80-0f09-11eb-89a9-65872bb7adf7.gif)

Reply:
:sweat_smile: